path_role.go

```diff
@@ -0,0 +1,433 @@
+// Constants for role specific data
+const (
+	DefaultTTLSeconds = 1800
+	MaxRoleNameLength = 50
+	// Increasing this above this limit might require implementing
+	// client-side paging in the filterGroupMembership API
+	MaxOCIDsPerRole = 100
+)
+
+func pathRole(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: "role/" + framework.GenericNameRegex("role"),
+		Fields: map[string]*framework.FieldSchema{
+			"role": {
+				Type:        framework.TypeString,
+				Description: "Name of the role.",
+			},
+			"description": {
+				Type:        framework.TypeString,
+				Description: `A description of the role.`,
+			},
+			"add_ocid_list": {
+				Type:        framework.TypeCommaStringSlice,
+				Description: `A comma separated list of OCIDs to add.`,
+			},
+			"remove_ocid_list": {
+				Type:        framework.TypeCommaStringSlice,
+				Description: `A comma separated list of OCIDs to remove. Is applicable only for the UPDATE operation.`,
+			},
+			"ttl": {
+				Type:        framework.TypeInt,
+				Default:     0,
+				Description: `Duration in seconds after which the issued token should expire. Defaults to 1800 seconds.`,
+			},
+			"add_policy_list": {
+				Type:        framework.TypeCommaStringSlice,
+				Default:     "default",
+				Description: "A list of Policies to be set on tokens issued using this role.",
+			},
+			"remove_policy_list": {
+				Type:        framework.TypeCommaStringSlice,
+				Default:     "default",
+				Description: "A list of Policies to be remove from the role. Is applicable only for the UPDATE operation.",
+			},
+		},
+
+		ExistenceCheck: b.pathRoleExistenceCheck,
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.CreateOperation: b.pathRoleCreate,
+			logical.UpdateOperation: b.pathRoleUpdate,
+			logical.ReadOperation:   b.pathRoleRead,
+			logical.DeleteOperation: b.pathRoleDelete,
+		},
+
+		HelpSynopsis:    pathRoleSyn,
+		HelpDescription: pathRoleDesc,
+	}
+}
+
+func pathListRoles(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: "role/?",
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ListOperation: b.pathRoleList,
+		},
+
+		HelpSynopsis:    pathListRolesHelpSyn,
+		HelpDescription: pathListRolesHelpDesc,
+	}
+}
+
+// Establishes dichotomy of request operation between CreateOperation and UpdateOperation.
+// Returning 'true' forces an UpdateOperation, CreateOperation otherwise.
+func (b *backend) pathRoleExistenceCheck(ctx context.Context, req *logical.Request, data *framework.FieldData) (bool, error) {
+	entry, err := b.lockedOCIRole(ctx, req.Storage, strings.ToLower(data.Get("role").(string)))
+	if err != nil {
+		return false, err
+	}
+	return entry != nil, nil
+}
+
+// lockedOCIRole returns the properties set on the given role. This method
+// acquires the read lock before reading the role from the storage.
+func (b *backend) lockedOCIRole(ctx context.Context, s logical.Storage, roleName string) (*OCIRoleEntry, error) {
```

Comment:
Why is a locked variant of this and the Set call necessary?

Reply:
Hi Jim, the locked variant protects the role entry. I am following the same conventions in other Auth plugins in Vault. All of them protect the get/set & delete in a similar fashion using a Mutex.

Reply:
In this case, the role entry is simply been read from or written to storage, which is already an atomic operation (with respect to the entry). There is no shared memory that we're trying to protect. Do you have an example of the pattern your copying, as I can look and see what the difference is and why they're using a Mutex.

Reply:
Hi Jim, the pattern is similar to other plugins in Vault like:

AWS auth:
(Line 259 and other places)
https://github.com/hashicorp/vault/blob/master/builtin/credential/aws/path_role.go

Kubernetes auth:
(Line 108 and in other places)
https://github.com/hashicorp/vault-plugin-auth-kubernetes/blob/master/path_role.go

Reply:
Vault makes no guarantees about order of requests. And I mentioned earlier, a single Storage operation doesn't need a lock. You'll see locks around operations in a couple of cases:

* There is shared memory involved, e.g. a cached value or something on the backend object that is involved.
* The API operation is multi-step and the consistency matters. The AWS read, in addition to updating some cached data, is sometimes doing a role upgrade (which is a Put). To avoid weird data races from other requests between the Get and (possible) Put, they lock the whole read API operation. 

I checked with the author of the Kubernetes example and he thought that the lock probably wasn't needed. I can see how locking a List operation _might_ improve the consistency of a List result if there are Puts/Deletes while the storage engine is building the list, but it would be a pretty special use case that relied on that, IMHO.

So far I've don't think the OCI code has the more complicated flows.

Reply:
Thanks for the detailed explanation Jim. I have removed the locks from the path_role code.